mtt-management/src/mtt_types.rs

```diff
@@ -0,0 +1,446 @@
+    /// 16K regions of 4KB page sizes. The value is the PFN of the 4KB page
+    /// that contains 16K entries of 2-bit values that indicate whether the
+    /// address is confidential.
+    MttL1Dir(u64),
+    /// The 64-MB range containing the address has been partitioned into
+    /// 2MB regions. The 32-bit value is a bit mask that indicates whether
+    /// the 2MB region is non-confidential (bit value: 0) or confidential (bit value: 1).
+    Mtt2MBPages(u64),
+    /// The L2 entry corresponding to the address is invalid.
+    /// This should be treated as an invalid operational state.
+    MttInvalidL2,
+}
+
+/// Enumeration of the defined MTT L1 entry types.
+/// The bit encoding is as follows:
+/// 00b: The 4K region is non-confidential
+/// 01b: The 4K region is confidential
+/// 11b: Invalid encoding
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum MttL1EntryType {
+    /// The address is mapped to non-confidential 4K page.
+    MttNonConfidential4K = 0,
+    /// The address is mapped to a confidential 4K page.
+    MttConfidential4K = 1,
+    /// The L1 entry corresponding to the address is invalid.
+    /// This should be treated as an invalid operational state.
+    MttInvalidL1,
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+/// API-related enums for non-confidential and confidential types
+pub enum MttMemoryType {
+    /// A non-confidential memory range
+    NonConfidential = 0,
+    /// A confidential memory range
+    Confidential = 1,
+}
+
+// Bits 33:0
+const MTT_L2_INFO_MASK: u64 = 0x0003_ffff_ffff;
+
+use MttL2EntryType::*;
+impl MttL2EntryType {
+    fn from_raw(value: u64) -> MttL2EntryType {
+        // Bits 63:35 must be zero
+        if (value & 0xffff_fff0_0000_0000) != 0 {
+            return MttInvalidL2;
+        }
+
+        // Bits 35:34 are the type
+        let entry_type = (value & 0x000c_0000_0000) >> 34;
+        match entry_type {
+            // INFO must be 0 for 1GB entries
+            0 if (value & MTT_L2_INFO_MASK) == 0 => MttNonConfidential1G,
+            1 if (value & MTT_L2_INFO_MASK) == 0 => MttConfidential1G,
+            2 => MttL1Dir(value & MTT_L2_INFO_MASK),
+            3 => Mtt2MBPages(value & MTT_L2_INFO_MASK),
+            _ => MttInvalidL2,
+        }
+    }
+
+    fn raw(&self) -> u64 {
+        match *self {
+            MttNonConfidential1G => 0,
+            MttConfidential1G => 1u64 << 34,
+            MttL1Dir(v) => (2u64 << 34) | v,
+            Mtt2MBPages(v) => (3u64 << 34) | v,
+            _ => unreachable!(),
+        }
+    }
+
+    fn confidential_64_mb_range() -> MttL2EntryType {
+        Mtt2MBPages(0xffff_ffff)
+    }
+
+    fn non_confidential_64_mb_range() -> MttL2EntryType {
+        Mtt2MBPages(0)
+    }
+
+    fn mtt_l1_dir(mtt_l1_page: SupervisorPageAddr) -> MttL2EntryType {
+        MttL1Dir(mtt_l1_page.bits() >> 12)
+    }
+}
+
+use MttL1EntryType::*;
+impl MttL1EntryType {
+    fn from_raw(value: u64) -> MttL1EntryType {
```

Comment:
Same discourse of above. Return `Result` instead of defining `MttInvalidL1`?

Reply:
See above...it might just be better to immediately panic.